k2/python/tests/rnnt_loss_test.py

```diff
@@ -488,23 +546,211 @@ def test_rnnt_loss_pruned(self):
+
+                logits = am.unsqueeze(2) + lm.unsqueeze(1)
+                logits = logits.float()
+
+                # nonlinear transform
+                logits = torch.sigmoid(logits)
+
+                k2_loss = k2.rnnt_loss(
+                    logits=logits,
+                    symbols=symbols,
+                    termination_symbol=terminal_symbol,
+                    boundary=boundary,
+                    modified=modified,
+                    reduction="none",
+                )
+
+                print(
+                    f"Unpruned rnnt loss with modified {modified} : {k2_loss}"
+                )
+
+                # pruning
+                k2_simple_loss, (px_grad, py_grad) = k2.rnnt_loss_simple(
+                    lm=lm,
+                    am=am,
+                    symbols=symbols,
+                    termination_symbol=terminal_symbol,
+                    boundary=boundary,
+                    modified=modified,
+                    return_grad=True,
+                    reduction="none",
+                )
+
+                S0 = 2
+                if modified:
+                    S0 = 1
+
+                for r in range(S0, S + 2):
+                    ranges = k2.get_rnnt_prune_ranges(
+                        px_grad=px_grad,
+                        py_grad=py_grad,
+                        boundary=boundary,
+                        s_range=r,
+                    )
+                    # (B, T, r, C)
+                    pruned_am, pruned_lm = k2.do_rnnt_pruning(
+                        am=am, lm=lm, ranges=ranges
+                    )
 
-                    t_prob_p = am_p + lm_p
+                    logits = pruned_am + pruned_lm
 
                     # nonlinear transform
-                    t_prob_p = torch.sigmoid(t_prob_p)
+                    logits = torch.sigmoid(logits)
 
                     pruned_loss = k2.rnnt_loss_pruned(
-                        logits=t_prob_p,
+                        logits=logits,
                         symbols=symbols,
                         ranges=ranges,
                         termination_symbol=terminal_symbol,
                         boundary=boundary,
                         modified=modified,
                         reduction="none",
                     )
-                    print(f"pruning loss with range {r} : {pruned_loss}")
+                    print(f"Pruned loss with range {r} : {pruned_loss}")
+
+    # Test more exact pruning bounds.
+    # In our previous versions we use a less exact method to generate
+    # pruning bounds which is different from the method we
+    # publish in our paper(https://arxiv.org/pdf/2206.13236.pdf).
+    # This tests the difference between these two methods.
+    # We won't do any assertion in this test, just printing out the losses,
+    # because we can not 100% sure that the new method is better than the old
+    # one all the time, both of them are local optimal bounds.
+    def test_prune_ranges(self):
```

Comment:
I tested the two pruning bounds here, first, I generated a mask according to the ranges, and filled the logits  with 0.0 at the positions outof ranges.  then, I computed the loss of the modifed logits.  If the new pruning bounds is better than the old one, the loss should be smaller.  From my experiments (both on real data and random generated data), most of the time, the loss of the logit pruned with new bounds is smaller than the one pruned with old bounds. But it can not 100% gurantee it. If I did something wrong, please LMK @danpovey, thanks!